nat-lab/tests/test_client_basic_stun.py

```diff
@@ -24,3 +30,70 @@ async def test_client_basic_stun(connection_tag: ConnectionTag, public_ip: str)
     async with new_connection_by_tag(connection_tag) as connection:
         ip = await testing.wait_long(stun.get(connection, config.STUN_SERVER))
         assert ip == public_ip, f"wrong public ip for the client {ip}"
+
+@pytest.mark.asyncio
+# Regression test for LLT-4306
+async def test_infinite_loop_in_stun_when_nonet() -> None:
+    async with AsyncExitStack() as exit_stack:
+        api = API()
+
+        (alpha, beta) = api.default_config_two_nodes()
+        alpha_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_1)
+        )
+        beta_connection = await exit_stack.enter_async_context(
+            new_connection_by_tag(ConnectionTag.DOCKER_CONE_CLIENT_2)
+        )
+
+        alpha_client = await exit_stack.enter_async_context(
+            Client(
+                alpha_connection,
+                alpha,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(
+                api.get_meshmap(alpha.id),
+            )
+        )
+        beta_client = await exit_stack.enter_async_context(
+            Client(
+                beta_connection,
+                beta,
+                AdapterType.BoringTun,
+                telio_features=TelioFeatures(direct=Direct(providers=["stun"])),
+            ).run_meshnet(api.get_meshmap(beta.id))
+        )
+
+        await testing.wait_defined(
+            asyncio.gather(
+                alpha_client.wait_for_state_peer(
+                    beta.public_key, [State.Connected], [PathType.Direct]
+                ),
+                beta_client.wait_for_state_peer(
+                    alpha.public_key, [State.Connected], [PathType.Direct]
+                ),
+            ),
+            60,
+        )
+
+        for server in config.DERP_SERVERS:
+            await exit_stack.enter_async_context(
+                alpha_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
+            await exit_stack.enter_async_context(
+                beta_client.get_router().break_udp_conn_to_host(str(server["ipv4"]))
+            )
+
+        # 3478 and 3479 are STUN ports in natlab containers
+        tcpdump = await exit_stack.enter_async_context(
+            alpha_connection.create_process(
+                ["tcpdump", "-i", "any", "(", "port", "3478", "or", "3479", ")"]
```

Comment:
if you reject outgoing packets, do they not show up on tcpdump?

Reply:
They're not really rejected, the firewall just responds with `icmp-host-unreachable`. And it does indeed show up in tcpdump

Reply:
The point here of checking `tcpdump` is that the bug would make the code spam the STUN server in an infinite loop. The requests are supposed to show up in `tcpdump` (and they do), there just shouldn't be that many requests

Reply:
riiight, thanks for explanation 👍